File: Java/Programmers/ship_box.java
package Programmers;

import java.util.Stack;

public class ship_box {
    public int solution(int[] order) {
        Stack<Integer> stack = new Stack<>();
        int box = 1;
        int target_idx = 0;

        while (target_idx < order.length){
            if (order[target_idx] ==  box){
                target_idx ++;
                box ++;
            } else if (stack.size() > 0 && order[target_idx] == stack.peek()){
                target_idx ++;
                stack.pop();
            } else if (box < order.length){
                stack.add(box);
                box ++;
            } else {
                break;
            }
        }
        return target_idx;
    }
}

File: Python/Programmers/ship_box.py
def solution(order):
    stack = []
    box = 1
    target_idx = 0
    
    while target_idx < len(order):
        if order[target_idx] == box:
            target_idx += 1
            box += 1
        elif len(stack) > 0 and order[target_idx] == stack[-1]:
            target_idx += 1
            stack.pop()
        elif box <= len(order):
            stack.append(box)
            box += 1
        else:
            break
        
    return target_idx